### PR TITLE
Machine-independent SD card driver

### PR DESCRIPTION
**Pull request by mohrcore**
Please, note that the changes in the emmc interface are also a part of #1101. Once merged they should be not visible here.

### CODE REVIEW

File: sys/drv/sd.c [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+
+    c++;
+    buf += 128;
+  }
+
+  return num * DEFAULT_BLKSIZE;
+}
+
+static inline int sd_check_uio(uio_t *uio) {
+  if (uio->uio_resid & (DEFAULT_BLKSIZE - 1))
```
Comment (MichalBlk) [inline sys/drv/sd.c:310]: Use `is_aligned` (sys/mimiker.h).

File: sys/drv/sd.c [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+      return 0;
+
+    c++;
+    buf += 128;
+  }
+
+  return num * DEFAULT_BLKSIZE;
+}
+
+static inline int sd_check_uio(uio_t *uio) {
```
Comment (MichalBlk) [inline sys/drv/sd.c:309]: This function imposes a somehow strange constraint. Why can't I write 513 bytes to the SD card? In USB mass storage driver, in cases like this, I read the context of the block I'm about to overwrite only partially, and I write the unchanged part as it is. Why don't you  do that?

File: sys/drv/sd.c [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+}
+
+typedef int (*sd_transfer_t)(device_t *dev, uint32_t lba, void *buffer,
+                             uint32_t num);
+
+static int sd_dop_uio(devnode_t *d, uio_t *uio) {
+  device_t *dev = d->data;
+  sd_state_t *state = (sd_state_t *)dev->state;
+
+  if (sd_check_uio(uio))
```
Comment (MichalBlk) [inline sys/drv/sd.c:324]: Please use error codes (sys/errno.h).

File: sys/drv/sd.c [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+
+static devops_t sd_devops = {
+  .d_read = sd_dop_uio,
+  .d_write = sd_dop_uio,
+};
+
+static int sd_attach(device_t *dev) {
+  sd_state_t *state = (sd_state_t *)dev->state;
+
+  state->block_buf = kmalloc(M_DEV, SD_KERNEL_BLOCKS * DEFAULT_BLKSIZE, 0);
```
Comment (MichalBlk) [inline sys/drv/sd.c:352]: `M_WAITOK` is preferable (its value is 0).

File: sys/drv/sd.c [hunk 5]
```diff
@@ -0,0 +1,371 @@
...
+        return 0;
+    }
+    if (emmc_write(dev, buf, DEFAULT_BLKSIZE, NULL)) {
+      klog("e.MMC write failed");
+      return 0;
+    }
+    if (emmc_wait(dev, EMMC_I_DATA_DONE))
+      return 0;
+
+    c++;
```
Comment (MichalBlk) [inline sys/drv/sd.c:302]: Please move it to a for statement.

File: sys/drv/sd.c [hunk 6]
```diff
@@ -0,0 +1,371 @@
...
+  }
+  while (c < num) {
+    if (!(state->props & SD_SUPP_BLKCNT)) {
+      emmc_send_cmd(dev, EMMC_CMD(WRITE_BLOCK), lba + c, NULL);
+      if (emmc_wait(dev, EMMC_I_WRITE_READY))
+        return 0;
+    }
+    if (emmc_write(dev, buf, DEFAULT_BLKSIZE, NULL)) {
+      klog("e.MMC write failed");
+      return 0;
```
Comment (MichalBlk) [inline sys/drv/sd.c:297]: Why do you return 0 in case of error?

File: sys/drv/sd.c [hunk 7]
```diff
@@ -0,0 +1,371 @@
...
+
+  uint32_t lba = uio->uio_offset / DEFAULT_BLKSIZE;
+  uint32_t blk_cnt = uio->uio_resid / DEFAULT_BLKSIZE;
+
+  sd_transfer_t transfer = uio->uio_op == UIO_READ ? sd_read_blk : sd_write_blk;
+
+  for (uint32_t blocks_left = blk_cnt; blocks_left;
+       blocks_left -= SD_KERNEL_BLOCKS) {
+    uint32_t num = min(blocks_left, (uint32_t)SD_KERNEL_BLOCKS);
+    transfer(dev, lba, state->block_buf, num);
```
Comment (MichalBlk) [inline sys/drv/sd.c:335]: Why don't you check/use the returned value!?

File: sys/drv/sd.c [hunk 8]
```diff
@@ -0,0 +1,371 @@
...
+
+  return 0;
+}
+
+static driver_t sd_block_device_driver = {
+  .desc = "SD(SC/HC) block device driver",
+  .size = sizeof(sd_state_t),
+  .probe = sd_probe,
+  .attach = sd_attach,
+  .pass = SECOND_PASS,
```
Comment (MichalBlk) [inline sys/drv/sd.c:367]: `pass` should be placed immediately under the `size` (this is a convention suggested by @cahirwpz).

File: sys/drv/sd.c [hunk 9]
```diff
@@ -0,0 +1,371 @@
...
+  .interfaces = {},
```
Comment (MichalBlk) [inline sys/drv/sd.c:368]: If it's nothing, why do you mention it? (You don't mention other fields like `detach`.)

**Review by MichalBlk — CHANGES_REQUESTED** [review]
Please create a separate PR with changes to the header file as this is something common to both your PRs and can be merged quickly.

File: sys/drv/sd.c (hunk 2 above)
Comment (MichalBlk) [reply sys/drv/sd.c:309]: Ok. Now I understand that it is the buffering layer which should take care of that.

File: include/dev/emmc.h [hunk 10]
```diff
@@ -234,7 +237,8 @@ static inline int emmc_read(device_t *dev, void *buf, size_t len, size_t *n) {
  * \param buf pointer to where the data should be read from
  * \param len data length
  * \param n pointer for the number of written bytes in or NULL
- * \return 0 on success
+ * \return 0 on success, EBUSY if device is busy, EINVAL on incorrect data
+ * length (not a multiple of block size)
```
[the remark this answers is not in the document]
Comment (mohrcore) [reply include/dev/emmc.h:241]: I think the comment is outdated at this point. Will fix it soon.

File: sys/drv/sd.c [hunk 11]
```diff
@@ -0,0 +1,371 @@
...
+  if (emmc_wait(dev, EMMC_I_DATA_DONE)) {
+    klog("SD card timed out when waiting for end of transmission");
+    return -1;
+  }
+  if (of != 64) {
+    klog("Failed to read SD Card's SCR");
+    return -1;
+  }
+
+  if (scr[0] & SCR_SUPP_SET_BLKCNT)
```
[the remark this answers is not in the document]
Comment (mohrcore) [reply sys/drv/sd.c:182]: **SCR** stands for "SD Configuration Register". This abbreviation is commonly used in the specs, but sure I can make a little comment.

File: sys/drv/sd.c [hunk 12]
```diff
@@ -0,0 +1,371 @@
...
+  EMMC_FMASK48((r), SD_ACMD41_RESP_CCS_OFFSET, SD_ACMD41_RESP_CCS_WIDTH)
+
+// SCR flags
+#define SCR_SD_BUS_WIDTH_4 0x00000400
+#define SCR_SUPP_SET_BLKCNT 0x02000000
+
+#define SD_BUSWIDTH_1 0x00
+#define SD_BUSWIDTH_4 0x02
+
+typedef enum sd_props {
```
[the remark this answers is not in the document]
Comment (mohrcore) [reply sys/drv/sd.c:84]: The semantics were meant to be different. `emmc_prop_id` is meant to be used as an identifier for a single property. `sd_props` were suppoused to be flags, but I forgot to give explicit values to its variants, which also caused a problem with `CCS` support detection. This has been fixed in my recent commit.

File: sys/drv/sd.c [hunk 13]
```diff
@@ -0,0 +1,371 @@
...
+  }
+  emmc_send_cmd(dev, SD_CMD_SET_IF_COND, propv << 8 | chkpat, &response);
+  if (SD_R7_CHKPAT(&response) != chkpat) {
+    klog("SD 2.0 voltage supply is mismatched, or the card is at Version 1.x");
+    return -1;
+  }
+  trial_cnt = 120;
+  /* Counter-intuitively, the busy bit is set ot 0 if the card is not ready */
+  SD_ACMD41_RESP_SET_BUSY(&response, 0);
+  while (trial_cnt & ~SD_ACMD41_RESP_READ_BUSY(&response)) {
```
[the remark this answers is not in the document]
Comment (mohrcore) [reply sys/drv/sd.c:139]: How would you check if `SD_ACMD41_RESP_READ_BUSY(&response)` succeeded on any trail? I can make a `success` variable, but that would just replace `trial_cnt` with a different variable of a scope that goes outside of the loop. I could also check the condition in the body of the loop ad use `goto` if it's true, but it's not the most elegant solution either.

File: sys/drv/sd.c [hunk 14]
```diff
@@ -0,0 +1,371 @@
...
+    }
+    emmc_send_cmd(dev, SD_CMD_SEND_OP_COND, SD_ACMD41_SD2_0_POLLRDY_ARG1,
+                  &response);
+  }
+  if (SD_ACMD41_RESP_READ_CCS(&response))
+    state->props |= SD_SUPP_CCS;
+  klog("e.MMC device detected as %s",
+       (state->props & SD_SUPP_CCS) ? high_cap_str : standard_cap_str);
+
+  /* Let's assume there's no voltage switching needed */
```
[the remark this answers is not in the document]
Comment (mohrcore) [reply sys/drv/sd.c:152]: Could you remind me what `XXX` stands for? I genuinely don't remember.

File: sys/drv/sd.c [hunk 15]
```diff
@@ -0,0 +1,371 @@
...
+
+/* The custom R7 response is handled just like R1 response, but has different
+ * bitfields, same goes for R6 */
+#define SDRESP_R7 EMMCRESP_R1
+#define SD_R7_CHKPAT(resp) EMMC_FMASK48(resp, 0, 8)
+
+#define SDRESP_R6 EMMCRESP_R1
+#define SD_R6_RCA(resp) EMMC_FMASK48(resp, 16, 16)
+
+/* Custom commands */
```
[the remark this answers is not in the document]
Comment (mohrcore) [reply sys/drv/sd.c]: I moved some stuff to `include/dev/sd.h`.

File: sys/drv/sd.c [hunk 16]
```diff
@@ -0,0 +1,371 @@
...
+#include <sys/rman.h>
+#include <sys/klog.h>
+#include <sys/vnode.h>
+#include <sys/devfs.h>
+
+/* Must be a power of two */
+#define DEFAULT_BLKSIZE 512
+#define SD_KERNEL_BLOCKS 4 /* Number of blocks buffered in kernel memory */
+
+static driver_t sd_block_device_driver;
```
[the remark this answers is not in the document]
Comment (mohrcore) [reply sys/drv/sd.c]: Removed.

File: sys/drv/sd.c [hunk 17]
```diff
@@ -0,0 +1,371 @@
...
+      buf += 128;
+    }
+  }
+  if (num > 1 && !sup_blkcnt && (state->props & SD_SUPP_CCS))
+    emmc_send_cmd(dev, EMMC_CMD(STOP_TRANSMISSION), 0, NULL);
+  return num * DEFAULT_BLKSIZE;
+}
+
+/**
+ * Write a block to the sd card
```
[the remark this answers is not in the document]
Comment (mohrcore) [reply sys/drv/sd.c]: Fixed.

File: sys/drv/sd.c [hunk 18]
```diff
@@ -0,0 +1,371 @@
...
+#include <sys/mimiker.h>
+#include <dev/emmc.h>
+#include <sys/device.h>
+#include <sys/devclass.h>
+#include <sys/rman.h>
+#include <sys/klog.h>
+#include <sys/vnode.h>
+#include <sys/devfs.h>
+
+/* Must be a power of two */
```
[the remark this answers is not in the document]
Comment (mohrcore) [reply sys/drv/sd.c]: I'm not sure how exactly could I assert this statically.

File: sys/drv/sd.c [hunk 19]
```diff
@@ -0,0 +1,371 @@
...
+  (emmc_cmd_t) {                                                               \
+    .cmd_idx = 3, .flags = 0, .exp_resp = SDRESP_R6,                           \
+  }
+
+#define SD_CMD_SET_BUS_WIDTH                                                   \
+  (emmc_cmd_t) {                                                               \
+    .cmd_idx = 6, .flags = EMMC_F_APP, .exp_resp = EMMCRESP_R1,                \
+  }
+
+#define SD_ACMD41_SD2_0_POLLRDY_ARG1 0x51ff8000
```
[the remark this answers is not in the document]
Comment (mohrcore) [reply sys/drv/sd.c]: There's a new comment now.

File: sys/drv/sd.c [hunk 20]
```diff
@@ -0,0 +1,371 @@
...
+    klog("Unable to determine whether CMD8 responds.");
+    return -1;
+  }
+  emmc_send_cmd(dev, SD_CMD_SET_IF_COND, propv << 8 | chkpat, &response);
+  if (SD_R7_CHKPAT(&response) != chkpat) {
+    klog("SD 2.0 voltage supply is mismatched, or the card is at Version 1.x");
+    return -1;
+  }
+  trial_cnt = 120;
+  /* Counter-intuitively, the busy bit is set ot 0 if the card is not ready */
```
[the remark this answers is not in the document]
Comment (mohrcore) [reply sys/drv/sd.c:77]: Fixed.

File: sys/drv/sd.c [hunk 21]
```diff
@@ -0,0 +1,371 @@
...
+
+  /* The routine below is based on SD Specifications: Part 1 Physical Layer
+   * Simplified Specification, Version 6.0. See: page 30 */
+
+  klog("Attaching SD/SDHC block device interface...");
+
+  emmc_send_cmd(dev, EMMC_CMD(GO_IDLE), 0, NULL);
+  if (emmc_get_prop(dev, EMMC_PROP_R_VOLTAGE_SUPPLY, &propv)) {
+    klog("Unable to determine e.MMC controller's voltage supply.");
+    return -1;
```
[the remark this answers is not in the document]
Comment (mohrcore) [reply sys/drv/sd.c]: I introduced error code into my code.

File: sys/drv/sd.c [hunk 22]
```diff
@@ -0,0 +1,371 @@
...
+  if (emmc_set_prop(dev, EMMC_PROP_RW_RESP_LOW, chkpat - 1)) {
+    klog("Unable to determine whether CMD8 responds.");
+    return -1;
+  }
+  emmc_send_cmd(dev, SD_CMD_SET_IF_COND, propv << 8 | chkpat, &response);
+  if (SD_R7_CHKPAT(&response) != chkpat) {
+    klog("SD 2.0 voltage supply is mismatched, or the card is at Version 1.x");
+    return -1;
+  }
+  trial_cnt = 120;
```
[the remark this answers is not in the document]
Comment (mohrcore) [reply sys/drv/sd.c]: Added a comment about that.

File: sys/drv/sd.c [hunk 23]
```diff
@@ -0,0 +1,371 @@
...
+
+  emmc_send_cmd(dev, EMMC_CMD(SELECT_CARD), rca << 16, NULL);
+  emmc_set_prop(dev, EMMC_PROP_RW_BLKSIZE, 8);
+  emmc_set_prop(dev, EMMC_PROP_RW_BLKCNT, 1);
+  emmc_send_cmd(dev, SD_CMD_SEND_SCR, 0, NULL);
+  if (emmc_wait(dev, EMMC_I_READ_READY)) {
+    klog("SD card timed out when waiting for data (SD_CMD_SEND_SCR)");
+    return -1;
+  }
+  emmc_read(dev, scr, 64, &of);
```
[the remark this answers is not in the document]
Comment (mohrcore) [reply sys/drv/sd.c]: Fixed.

File: sys/drv/sd.c [hunk 24]
```diff
@@ -0,0 +1,371 @@
...
+
+  klog("Card's feature support:\n* CCS: %s\n* SET_BLOCK_COUNT: %s",
+       (state->props & SD_SUPP_CCS) ? "YES" : "NO",
+       (state->props & SD_SUPP_BLKCNT) ? "YES" : "NO");
+
+  return 0;
+}
+
+/**
+ * read a block from sd card and return the number of bytes read
```
[the remark this answers is not in the document]
Comment (mohrcore) [reply sys/drv/sd.c]: Fixed.

File: sys/drv/sd.c [hunk 25]
```diff
@@ -0,0 +1,371 @@
...
+ * returns 0 on error.
+ */
+int sd_read_blk(device_t *dev, uint32_t lba, void *buffer, uint32_t num) {
```
[the remark this answers is not in the document]
Comment (mohrcore) [reply sys/drv/sd.c]: Fixed.

[the remark this answers is not in the document]
Comment (mohrcore) [reply sys/drv/sd.c]: I changed it so now it return error codes instead.

File: sys/drv/sd.c [hunk 26]
```diff
@@ -0,0 +1,371 @@
...
+      return 0;
+  } else {
+    for (uint32_t c = 0; c < num; c++) {
+      emmc_send_cmd(dev, EMMC_CMD(READ_BLOCK), lba + c, NULL);
+      r = emmc_read(dev, buf, 512, NULL);
+      if (r)
+        return 0;
+      if (emmc_wait(dev, EMMC_I_DATA_DONE))
+        return 0;
+      buf += 128;
```
[the remark this answers is not in the document]
Comment (mohrcore) [reply sys/drv/sd.c]: Done. I see no purpouse in doing the addition in the increment part of the for statement tho.

File: sys/drv/sd.c (hunk 17 above)
[the remark this answers is not in the document]
Comment (mohrcore) [reply sys/drv/sd.c]: Fixed.

File: sys/drv/sd.c [hunk 27]
```diff
@@ -0,0 +1,371 @@
...
+  if (num > 1 && !sup_blkcnt && (state->props & SD_SUPP_CCS))
+    emmc_send_cmd(dev, EMMC_CMD(STOP_TRANSMISSION), 0, NULL);
+  return num * DEFAULT_BLKSIZE;
+}
+
+/**
+ * Write a block to the sd card
+ * returns 0 on error
+ */
+int sd_write_blk(device_t *dev, uint32_t lba, void *buffer, uint32_t num) {
```
[the remark this answers is not in the document]
Comment (mohrcore) [reply sys/drv/sd.c]: Fixed.

File: sys/drv/sd.c (hunk 9 above)
Comment (mohrcore) [reply sys/drv/sd.c]: Fixed.

File: sys/drv/sd.c (hunk 7 above)
Comment (mohrcore) [reply sys/drv/sd.c]: Fixed.

File: sys/drv/sd.c (hunk 6 above)
Comment (mohrcore) [reply sys/drv/sd.c]: It's error codes now.

File: sys/drv/sd.c (hunk 5 above)
Comment (mohrcore) [reply sys/drv/sd.c]: Fixed.

File: sys/drv/sd.c (hunk 3 above)
Comment (mohrcore) [reply sys/drv/sd.c]: Now I do.

File: sys/drv/sd.c [hunk 28]
```diff
@@ -0,0 +1,371 @@
...
+    emmc_send_cmd(
+      dev, num == 1 ? EMMC_CMD(WRITE_BLOCK) : EMMC_CMD(WRITE_MULTIPLE_BLOCKS),
+      lba, NULL);
+    if (emmc_wait(dev, EMMC_I_WRITE_READY))
+      return 0;
+  } else {
+    emmc_set_prop(dev, EMMC_PROP_RW_BLKCNT, num);
+    emmc_set_prop(dev, EMMC_PROP_RW_BLKSIZE, DEFAULT_BLKSIZE);
+  }
+  while (c < num) {
```
[the remark this answers is not in the document]
Comment (mohrcore) [reply sys/drv/sd.c]: I changed the `while` loop into a `for` loop.

File: sys/drv/sd.c (hunk 8 above)
Comment (mohrcore) [reply sys/drv/sd.c]: Fixed.

File: sys/drv/sd.c [hunk 29]
```diff
@@ -0,0 +1,371 @@
...
+ */
+int sd_read_blk(device_t *dev, uint32_t lba, void *buffer, uint32_t num) {
+  assert(dev->driver == (driver_t *)&sd_block_device_driver);
+  sd_state_t *state = (sd_state_t *)dev->state;
+
+  if (num < 1)
+    return 0;
+
+  uint32_t *buf = (uint32_t *)buffer;
+  int r;
```
[the remark this answers is not in the document]
Comment (mohrcore) [reply sys/drv/sd.c]: Fixed.

File: sys/drv/sd.c [hunk 30]
```diff
@@ -0,0 +1,371 @@
...
+
+  return 0;
+}
+
+/**
+ * read a block from sd card and return the number of bytes read
+ * returns 0 on error.
+ */
+int sd_read_blk(device_t *dev, uint32_t lba, void *buffer, uint32_t num) {
+  assert(dev->driver == (driver_t *)&sd_block_device_driver);
```
[the remark this answers is not in the document]
Comment (mohrcore) [reply sys/drv/sd.c]: I removed those statements.

File: sys/drv/sd.c [hunk 31]
```diff
@@ -0,0 +1,371 @@
...
+  if (state->props & SD_SUPP_CCS) {
+    r = emmc_read(dev, buf, num * DEFAULT_BLKSIZE, NULL);
+    if (r)
+      return 0;
+    if (emmc_wait(dev, EMMC_I_DATA_DONE))
+      return 0;
+  } else {
+    for (uint32_t c = 0; c < num; c++) {
+      emmc_send_cmd(dev, EMMC_CMD(READ_BLOCK), lba + c, NULL);
+      r = emmc_read(dev, buf, 512, NULL);
```
[the remark this answers is not in the document]
Comment (mohrcore) [reply sys/drv/sd.c]: Fixed.

File: sys/drv/sd.c (hunk 14 above)
[the remark this answers is not in the document]
Comment (cahirwpz) [reply sys/drv/sd.c:131]: https://github.com/cahirwpz/mimiker/wiki/commenting

**Comment by mohrcore** [comment]
Requires https://github.com/cahirwpz/mimiker/pull/1331 to write/read at adresses other than 0.

**Comment by mohrcore** [comment]
Requires https://github.com/cahirwpz/mimiker/pull/1284

**Comment by mohrcore** [comment]
To test the driver: merge the required PRs, if still unmerged, then merge https://github.com/mohrcore/mimiker/tree/sdr_sdw, which contains simple userspace programs for reading from and writing to sd card. Note, this branch is not meant to be merge into the mainline. It's purpose is only to provide very primitive tools for testing reads and writes to the sd_card device as a temporary substitue for `dd`.

Example usage:

Write 13 bytes at byte address 100:
```
echo "Hello World!" | sdw 100 13
```

Read 13 bytes at byte address 100:
```
$ sdr 100 13
Hello World!
```

**Review by cahirwpz — APPROVED** [review]
Though the changes makes me feel uneasy I'll merge that and perhaps we'll return to this code later.

**Pull request merged**